solr/solrj/src/java/org/apache/solr/common/cloud/CollectionStateWatcher.java

```diff
@@ -0,0 +1,42 @@
+package org.apache.solr.common.cloud;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+import java.util.Set;
+
+/**
+ * Callback registered with {@link ZkStateReader#registerCollectionStateWatcher(String, CollectionStateWatcher)}
+ * and called whenever the collection state changes.
+ */
+public interface CollectionStateWatcher {
+
+  /**
+   * Called when the collection we are registered against has a change of state
+   *
+   * Note that, due to the way Zookeeper watchers are implemented, a single call may be
+   * the result of several state changes
+   *
+   * A watcher is unregistered after it has been called once.  To make a watcher persistent,
+   * implementors should re-register during this call.
+   *
+   * @param liveNodes       the set of live nodes
+   * @param collectionState the new collection state
+   */
+  void onStateChanged(Set<String> liveNodes, DocCollection collectionState);
+
+}
```

Comment:
I just want to toss out an idea here after looking at this some more.  I notice that CollectionStateWatcher and CollectionStatePredicate are nearly identical.  What would you think about combining the two into a single interface?

The signature could be e.g.:

bool stateChanged(liveNodes, collectionState)

In a watch context, the return value means "keep watching?".  So return true to reset the watcher and continue getting updates, or return false to stop watching.

In a predicate context, the return value means "keep waiting?".  So return true to keep waiting, or return false if you've finally seen what you were waiting for.

They'll both have the same semantic meaning either way.